### PR TITLE
feat: cni refactor for swift v2

### PR DESCRIPTION
**Pull request by jaer-tsun**
combining 4 CNI PRs from forked repo

**Reason for Change**:

CNI changes for Swift 2
- update contract when CNI RequestIPs from CNS to include 
> - MacAddress of secondary (multitenant) interface
> - NICType to define whether the IP config is InfraNIC (existing scenarios) or DelegatedVMNIC (swift 2)
> - SkipDefaultRoutes to specify which interface traffic will go through by default
> - Routes to configure on interface
- refactor existing endpoint creation flow to support multiple 'EndpointInfo' which represents an interface and consolidate into 'endpoint' struct which maps to pod
> - added SecondaryInterfaces to endpoint struct
- add SecondaryEndpointClient to configure secondary interface

**Issue Fixed**:

**Requirements**:

- [x] uses [conventional commit messages](https://www.conventionalcommits.org/)
  
- [ ] includes documentation
- [x] adds unit tests
- [x] relevant PR labels added

**Notes**:

### CODE REVIEW

File: cni/network/invoker_azure_test.go [hunk 1]
```diff
@@ -127,15 +135,14 @@ func TestAzureIPAMInvoker_Add(t *testing.T) {
...
-		want1   *cniTypesCurr.Result
 		wantErr bool
 	}{
 		{
 			name: "happy add ipv4",
 			fields: fields{
 				plugin: &mockDelegatePlugin{
 					add: add{
-						resultsIPv4: getResult("10.0.0.1/24"),
+						resultsIPv4: getSingleResult("10.0.0.1/24"),
```
Comment (behzad-mir) [inline cni/network/invoker_azure_test.go:145]: Can you move the subnet to a var at the beginning of the file?

File: cni/network/invoker_azure_test.go [hunk 2]
```diff
@@ -145,16 +152,16 @@ func TestAzureIPAMInvoker_Add(t *testing.T) {
 				nwCfg:        &cni.NetworkConfig{},
 				subnetPrefix: getCIDRNotationForAddress("10.0.0.0/24"),
 			},
-			want:    getResult("10.0.0.1/24")[0],
+			want:    getResult("10.0.0.1/24"),
```
Comment (behzad-mir) [inline cni/network/invoker_azure_test.go:155]: same thing

File: cni/network/invoker_cns.go [hunk 3]
```diff
@@ -21,11 +21,17 @@ import (
...
-	errEmptyCNIArgs    = errors.New("empty CNI cmd args not allowed")
-	errInvalidArgs     = errors.New("invalid arg(s)")
-	overlayGatewayV6IP = "fe80::1234:5678:9abc"
-	watcherPath        = "/var/run/azure-vnet/deleteIDs"
+	errEmptyCNIArgs          = errors.New("empty CNI cmd args not allowed")
+	errInvalidArgs           = errors.New("invalid arg(s)")
+	errInvalidDefaultRouting = errors.New("add result requires exactly one interface with default routes")
+	errInvalidGatewayIP      = errors.New("invalid gateway IP")
+	overlayGatewayV6IP       = "fe80::1234:5678:9abc"
+	watcherPath              = "/var/run/azure-vnet/deleteIDs"
```
Comment (behzad-mir) [inline cni/network/invoker_cns.go:34]: Do we need a path for Windows?

File: cni/network/invoker_cns.go [hunk 4]
```diff
@@ -352,3 +314,151 @@ func (invoker *CNSIPAMInvoker) Delete(address *net.IPNet, nwCfg *cni.NetworkConf
...
+
+func configureSecondaryAddResult(info *IPResultInfo, addResult *IPAMAddResult, podIPConfig *cns.IPSubnet) error {
+	ip, ipnet, err := podIPConfig.GetIPNet()
+	if ip == nil {
+		return errors.Wrap(err, "Unable to parse IP from response: "+info.podIPAddress+" with err %w")
+	}
+
+	macAddress, err := net.ParseMAC(info.macAddress)
+	if err != nil {
+		return errors.Wrap(err, "Invalid mac address")
```
Comment (vipul-21) [inline cni/network/invoker_cns.go:436]: Can we add the invalid mac address to the error too ?

File: cni/network/invoker_cns.go (hunk 3 above)
Comment (jaer-tsun) [reply cni/network/invoker_cns.go:34]: not supported atm

File: cni/network/invoker_cns.go (hunk 4 above)
Comment (jaer-tsun) [reply cni/network/invoker_cns.go:436]: ParseMac err already has it

Comment (tamilmani1989) [reply cni/network/invoker_cns.go:436]: do we have UT for this and how are we validating this error?

**Comment by tamilmani1989** [comment]
@jaer-tsun can we also validate this commit with release pipelines.. check with @jpayne3506 on this

File: cni/network/invoker_cns.go (hunk 4 above)
Comment (jaer-tsun) [reply cni/network/invoker_cns.go:436]: yeah we have UT to expect err for bad mac

File: cni/network/invoker_azure_test.go (hunk 1 above)
Comment (jaer-tsun) [reply cni/network/invoker_azure_test.go]: updated

File: cni/network/invoker_azure_test.go (hunk 2 above)
Comment (jaer-tsun) [reply cni/network/invoker_azure_test.go]: updated

**Review by tamilmani1989 — APPROVED** [review]

**Review by vipul-21 — APPROVED** [review]

**Review by behzad-mir — APPROVED** [review]

**Pull request merged**